src/lib.rs

```diff
@@ -343,6 +348,136 @@ impl Invariant {
         self.positions.get_all(&owner_id)
     }
 
+    pub async fn swap(
+        &mut self,
+        pool_key: PoolKey,
+        x_to_y: bool,
+        amount: TokenAmount,
+        by_amount_in: bool,
+        sqrt_price_limit: SqrtPrice,
+    ) -> Result<CalculateSwapResult, InvariantError> {
+        let caller = msg::source();
+        let program = exec::program_id();
+
+        let calculate_swap_result =
+            self.calculate_swap(pool_key, x_to_y, amount, by_amount_in, sqrt_price_limit)?;
+
+        let mut crossed_tick_indexes: Vec<i32> = vec![];
+
+        for tick in calculate_swap_result.ticks.iter() {
+            crossed_tick_indexes.push(tick.index);
+        }
+
+        if !crossed_tick_indexes.is_empty() {
```

Comment:
Consider emitting the cross-tick event after checking the availability of gas (if the event is emitted in case of failure as well - not sure about it)